techlibs/intel_alm/synth_intel_alm.cc

```diff
@@ -0,0 +1,231 @@
+				nobram = true;
+				continue;
+			}
+			if (args[argidx] == "-noflatten") {
+				flatten = false;
+				continue;
+			}
+			break;
+		}
+		extra_args(args, argidx, design);
+
+		if (!design->full_selection())
+			log_cmd_error("This command only operates on fully selected designs!\n");
+
+		if (family_opt == "cyclonev") {
+			bram_type = "m10k";
+		} else if (family_opt == "cyclone10gx") {
+			bram_type = "m20k";
+		} else {
+			log_cmd_error("Invalid or no family specified: '%s'\n", family_opt.c_str());
+		}
+
+		log_header(design, "Executing SYNTH_INTEL_ALM pass.\n");
+		log_push();
+
+		run_script(design, run_from, run_to);
+
+		log_pop();
+	}
+
+	void script() YS_OVERRIDE
+	{
+		if (check_label("begin")) {
+			run(stringf("read_verilog -sv -lib +/intel/%s/cells_sim.v", family_opt.c_str()));
+			run("read_verilog -lib +/intel_alm/common/alm_sim.v");
+			run(stringf("read_verilog -lib -D %s +/intel_alm/common/dff_sim.v", family_opt.c_str()));
+
+			// Misc and common cells
+			run("read_verilog -lib +/intel/common/altpll_bb.v");
+			run("read_verilog -lib +/intel_alm/common/megafunction_bb.v");
+			run(stringf("hierarchy -check %s", help_mode ? "-top <top>" : top_opt.c_str()));
+		}
+
+		if (flatten && check_label("flatten", "(unless -noflatten)")) {
+			run("proc");
+			run("flatten");
+			run("tribuf -logic");
+			run("deminout");
+		}
+
+		if (check_label("coarse")) {
+			run("synth -run coarse -lut 6");
+			run("techmap -map +/intel_alm/common/arith_alm_map.v");
+		}
+
+		if (!nobram && check_label("map_bram", "(skip if -nobram)")) {
+			run(stringf("memory_bram -rules +/intel_alm/common/bram_%s.txt", bram_type.c_str()));
+			run(stringf("techmap -map +/intel_alm/common/bram_%s_map.v", bram_type.c_str()));
+		}
+
+		if (!nolutram && check_label("map_lutram", "(skip if -nolutram)")) {
+			run("memory_bram -rules +/intel_alm/common/lutram_mlab.txt", "(for Cyclone V / Cyclone 10GX)");
+			run("techmap -map +/intel_alm/common/lutram_mlab_map.v", "(for Cyclone V / Cyclone 10GX)");
+		}
+
+		if (check_label("map_ffram")) {
+			run("memory_map");
+			run("opt -full");
+		}
+
+		if (check_label("map_ffs")) {
+			run("dff2dffe -direct-match $_DFF_*");
+			run("techmap -map +/techmap.v -map +/intel_alm/common/dff_map.v");
+			run("opt -full -undriven -mux_undef");
+			run("clean -purge");
+		}
+
+		if (check_label("map_luts")) {
+			run(stringf("abc9 -lut +/intel_alm/%s/abc9_lut.lut -box +/intel_alm/%s/abc9_box.box -W 200", family_opt.c_str(), family_opt.c_str()));
+			run("hilomap -singleton -hicell __MISTRAL_VCC Q -locell __MISTRAL_GND Q");
+			run("techmap -map +/intel_alm/common/alm_map.v");
+			run("opt -fast");
+			run("autoname");
+			run("clean");
+		}
+
+		if (check_label("check")) {
+			//run("hierarchy -check");
+			run("stat");
+			run("check");
+		}
+
+		if (check_label("vqm")) {
+			if (!vout_file.empty() || help_mode) {
+				run(stringf("techmap -map +/intel_alm/%s/quartus_rename.v", family_opt.c_str()));
```

Comment:
This will cause different synthesis results depending on whether edif/vqm output is enabled — perhaps use `design -save/-load`, or add another option?

Reply:
I don't see much point in picking one primitive list over another, and being intentionally leaky like this makes it easier to write tests to additionally check that the right Quartus primitives are produced.

Reply:
I added the `-quartus` option to produce a Quartus-compatible netlist.